src/shared/codelens/localLambdaRunner.ts

```diff
@@ -532,72 +537,58 @@ const getEnvironmentVariables = (config: HandlerConfig): SAMTemplateEnvironmentV
-    startInvokeMillis: number,
-    runtime: string,
-}): Promise<{ success: boolean }> {
-    const channelLogger = params.channeLogger
-    const logger = params.channeLogger.logger
+export interface AttachDebuggerContext {
+    debugConfig: DebugConfiguration
+    maxAttempts: number
+    channelLogger: Pick<ChannelLogger, 'info' | 'error' | 'logger'>
+    onStartDebugging?: typeof vscode.debug.startDebugging
+    onRecordAttachDebuggerMetric?(attachResult: boolean | undefined, attempts: number, attachResultDate: Date): void
+    onWillRetry(): Promise<void>
+}
+
+export async function attachDebugger(
+    {
+        onStartDebugging = vscode.debug.startDebugging,
+        ...params
+    }: AttachDebuggerContext
+): Promise<{ success: boolean }> {
+    const channelLogger = params.channelLogger
+    const logger = params.channelLogger.logger
     logger.debug(`localLambdaRunner.attachDebugger: startDebugging with debugConfig: ${JSON.stringify(
         params.debugConfig,
         undefined,
         2
     )}`)
 
-    let isDebuggerAttached: boolean | undefined = false
+    let isDebuggerAttached: boolean | undefined
     let numAttempts = 0
-    let retryDelay = 1000
-    let shouldRetry = false
-    const retryEnabled = false // Change this to enable retry
+
+    channelLogger.info(
+        'AWS.output.sam.local.attaching',
+        'Attaching debugger to SAM Application...',
+    )
+
     do {
-        channelLogger.info(
-            'AWS.output.sam.local.attaching',
-            'Attempt number {0} to attach debugger to SAM Application...',
-            String(numAttempts + 1)
-        )
-        isDebuggerAttached = await vscode.debug.startDebugging(undefined, params.debugConfig)
         numAttempts += 1
-        if (isDebuggerAttached === undefined) {
-            isDebuggerAttached = false
-            shouldRetry = numAttempts < MAX_DEBUGGER_ATTEMPTS
-        } else if (!isDebuggerAttached) {
-            retryDelay *= 2
-
-            shouldRetry = retryEnabled && (numAttempts < MAX_DEBUGGER_ATTEMPTS)
-            if (shouldRetry) {
-                const currTime = new Date()
-                recordDebugAttachResult({
-                    telemetryService: params.telemetryService,
-                    attachResult: isDebuggerAttached,
-                    attempts: numAttempts,
-                    duration: currTime.getTime() - params.startInvokeMillis,
-                    runtime: params.runtime,
-                })
+        isDebuggerAttached = await onStartDebugging(undefined, params.debugConfig)
+
+        if (params.onRecordAttachDebuggerMetric) {
+            params.onRecordAttachDebuggerMetric(isDebuggerAttached, numAttempts, new Date())
+        }
 
-                channelLogger.info(
-                    'AWS.output.sam.local.attach.retry',
-                    'Will try to attach debugger again in {0} seconds...',
-                    String(retryDelay / 1000)
+        if (isDebuggerAttached === undefined) {
```

Comment:
I'm not huge on taking a hard dependence on this `undefined` here since this appears to be a bug (thanks for filing, @awschristou !): https://github.com/Microsoft/vscode/issues/71565 ... let's make sure we keep an eye on this one.

If this does get fixed to report `false` on bad connections, then what should we do?

Reply:
We'll determine how widespread the impact is, and look for additional root causes at that time.

Reply:
@awschristou At that point, the impact would be widespread, as all users that:

1. Fail to attach on the first attempt; and
2. Have the updated version of VS Code (this should be most users due to auto-update)

will be blocked from local debugging. This is way too late to catch the issue. We need to at least need confirm that this is/isn't a bug before publishing this change.